examples/multiple-executors/main.tf

```diff
@@ -1,3 +1,6 @@
+data "google_project" "project" {
+}
```

Comment:
WDYT about moving this data source inside the modules? That way it would be completely opaque to the user, be it v3 or v4.

Reply:
> WDYT about moving this data source inside the modules? That way it would be completely opaque to the user, be it v3 or v4.

maybe not really worth doing it? For example, people can also just do

```tf
module "executor" {
  project = "i-am-a-gcp-project-id"
}
```

The example uses `data "google_project"` is more like a suggestion - you should probably do it this way (I remember terraform will validate if the reference `data_source` actually exists in remote?)

Reply:
But it could be confusing that the project setting only affects the IAM resource but none else 🤔 Currently, it's all controlled by the provider that's passed. That we would keep when we did the data source hoop

Reply:
I think the module assumes all resources are created in the same project, and you can't have `google_compute_autoscaler` and `google_compute_firewall` created in two different projects. Things just won't work.

How about

### Option 1

Document the fact that `project` has to be the same as the `project` in the provider the consumer configured.

### Option 2

Update all resources in the module to explicitly reference the provided `project` to make sure everything is indeed created within the same project

Reply:
> I think the module assumes all resources are created in the same project, and you can't have google_compute_autoscaler and google_compute_firewall created in two different projects. Things just won't work.

Correct! Hence we so far 100% relied on the terraform provider to pick the project solely based on _provider settings_.

Adding an *option 3* to the list: If we just internalized the data source that you added to the consumers currently, we would in fact still be implicitly relying on how the provider is configured, there can't be discrepancies. 

I think option 1 is a little dangerous, as some customers WILL miss that and then we need to go through some painful debugging process with them. 

Option 2 and 3 both seem valid to me, but to keep things simple (so we don't have to write some validator or something to make sure that the explicit project setting is never missed on any resource in here), and because it's one less thing to worry about for users of the module, I personally would go with option 3. I would approve either 2 or 3 though, so you make the final call here.